web/api/v1/api.go

```diff
@@ -538,14 +546,50 @@ func (api *API) labelValues(r *http.Request) (result apiFuncResult) {
 		q.Close()
 	}
 
-	vals, warnings, err := q.LabelValues(name)
+	var vals []string
+	var warnings storage.Warnings
+	if len(r.Form["match[]"]) > 0 {
+		// Get all series which match matchers.
+		var sets []storage.SeriesSet
+		for _, mset := range matcherSets {
+			s := q.Select(false, nil, mset...)
+			sets = append(sets, s)
+		}
+		vals, warnings, err = labelValuesByMatchers(sets, name)
```

Comment:
Can we add TODO to extend LabelValues and LabelNames with matches?

Reply:
I would also prefer if this was done on TSDB level. Last time we discussed this we decided against it, but I'm happy to have that discussion again (I realize that TSDB doesn't work that way today).